pkg/handler/v2/external_cluster/node.go

```diff
@@ -84,52 +84,56 @@ func ListNodesMetricsEndpoint(userInfoGetter provider.UserInfoGetter, projectPro
 		if err := req.Validate(); err != nil {
 			return nil, errors.NewBadRequest(err.Error())
 		}
-		nodeMetrics := make([]apiv1.NodeMetric, 0)
+		return getClusterNodesMetrics(ctx, userInfoGetter, projectProvider, privilegedProjectProvider, clusterProvider, privilegedClusterProvider, req.ProjectID, req.ClusterID)
+	}
+}
 
-		project, err := common.GetProject(ctx, userInfoGetter, projectProvider, privilegedProjectProvider, req.ProjectID, &provider.ProjectGetOptions{IncludeUninitialized: false})
+func getClusterNodesMetrics(ctx context.Context, userInfoGetter provider.UserInfoGetter, projectProvider provider.ProjectProvider, privilegedProjectProvider provider.PrivilegedProjectProvider, clusterProvider provider.ExternalClusterProvider, privilegedClusterProvider provider.PrivilegedExternalClusterProvider, projectID, clusterID string) ([]apiv1.NodeMetric, error) {
+	nodeMetrics := make([]apiv1.NodeMetric, 0)
+
+	project, err := common.GetProject(ctx, userInfoGetter, projectProvider, privilegedProjectProvider, projectID, &provider.ProjectGetOptions{IncludeUninitialized: false})
+	if err != nil {
+		return nil, common.KubernetesErrorToHTTPError(err)
+	}
+	cluster, err := getCluster(ctx, userInfoGetter, clusterProvider, privilegedClusterProvider, project.Name, clusterID)
+	if err != nil {
+		return nil, common.KubernetesErrorToHTTPError(err)
+	}
+
+	isMetricServer, err := clusterProvider.IsMetricServerAvailable(cluster)
+	if err != nil {
+		return nil, common.KubernetesErrorToHTTPError(err)
+	}
+
+	if isMetricServer {
+		client, err := clusterProvider.GetClient(cluster)
 		if err != nil {
 			return nil, common.KubernetesErrorToHTTPError(err)
 		}
-		cluster, err := getCluster(ctx, userInfoGetter, clusterProvider, privilegedClusterProvider, project.Name, req.ClusterID)
-		if err != nil {
-			return nil, common.KubernetesErrorToHTTPError(err)
+		nodes := &corev1.NodeList{}
+		if err := client.List(ctx, nodes); err != nil {
```

Comment:
Can we filter out nodes for the desired machinedeployment here before fetching the metrics like this https://github.com/kubermatic/kubermatic/blob/master/pkg/handler/v2/external_cluster/gke.go#L168

Reply:
this is a general method for all providers and also for the connected clusters from kubeconfig, so the whole implementation work with all provider. Adding filtering here will cause additional maintenance in the future when we will add a new provider. 
It's used in two places: one without filtering (get metrics for cluster nodes) and second when the filtering is done automatically without distinguishing the provider type (filter out MD nodes)